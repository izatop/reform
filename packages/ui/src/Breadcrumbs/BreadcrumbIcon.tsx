import {Icon} from "@reform/bulma";
import * as React from "react";

interface BreadcrumbIconProps {
    icon?: string | React.ReactElement;
}

export const BreadcrumbIcon: React.FC<BreadcrumbIconProps> = ({icon}) => {
    if (!icon) {
        return null;
    }

    if (typeof icon === "string") {
        return <Icon icon={icon}/>;
    }

    return icon;
};
