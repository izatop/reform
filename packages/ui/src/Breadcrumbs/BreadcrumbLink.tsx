import {Icon} from "@reform/bulma";
import * as React from "react";

interface BreadcrumbIconProps {
    icon?: string | React.ReactElement;
    title: string | React.ReactElement;
}

export const BreadcrumbLink: React.FC<BreadcrumbIconProps> = ({icon, title}) => {
    if (!icon) {
        return <>{title}</>;
    }

    if (typeof icon === "string") {
        return <><Icon icon={icon}/><span>{title}</span></>;
    }

    return <>{icon}<span>{title}</span></>;
};
