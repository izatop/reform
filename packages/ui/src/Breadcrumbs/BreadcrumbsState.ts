import {ReactElement, useEffect, useState} from "react";

export type BreadcrumbPath = [string, string, (string | ReactElement)?];
export type BreadcrumbsListener = (paths: BreadcrumbPath[]) => any;

export class BreadcrumbsState {
    #store: BreadcrumbPath[] = [];
    public readonly base: BreadcrumbPath[];
    readonly #listeners: BreadcrumbsListener[] = [];

    constructor(base: BreadcrumbPath[]) {
        this.base = base;
        this.#store = [...base];
    }

    public get paths() {
        return [...this.#store];
    }

    public mount(...path: BreadcrumbPath[]) {
        useEffect(() => {
            this.#store.push(...path);
            this.emit(this.paths);

            return () => {
                this.#store = [...this.base];
                this.emit(this.paths);
            };
        });
    }

    public usePaths() {
        const [state, setState] = useState(this.paths);
        useEffect(() => this.listen((paths) => setState(paths)));
        return state;
    }

    protected listen(listener: BreadcrumbsListener) {
        this.#listeners.push(listener);
        return () => {
            const index = this.#listeners.findIndex((fn) => listener === fn);
            this.#listeners.splice(index, 1);
        };
    }

    protected emit(paths: BreadcrumbPath[]) {
        requestAnimationFrame(() => this.#listeners.forEach((fn) => fn(paths)));
    }
}
