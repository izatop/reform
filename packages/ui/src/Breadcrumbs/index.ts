export * from "./BreadcrumbsProvider";
export * from "./BreadcrumbsState";
export * from "./Breadcrumbs";
export * from "./functions";
