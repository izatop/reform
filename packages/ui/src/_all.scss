@import "~@reform/bulma/sass/variables";
@import "./Date/all";
