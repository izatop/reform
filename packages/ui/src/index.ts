export * from "./Accesibility";
export * from "./Table";
