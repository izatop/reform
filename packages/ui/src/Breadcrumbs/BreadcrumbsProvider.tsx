import * as React from "react";
import {createContext, useMemo} from "react";
import {BreadcrumbPath, BreadcrumbsState} from "./BreadcrumbsState";

export interface BreadcrumbsProviderProps {
    base: BreadcrumbPath[];
}

export const BreadcrumbsProvider: React.FC<BreadcrumbsProviderProps> = ({children, base}) => {
    const breadcrumbs = useMemo(() => new BreadcrumbsState(base), [base]);
    return (
        <BreadcrumbsContext.Provider value={breadcrumbs}>
            {children}
        </BreadcrumbsContext.Provider>
    );
};

export const BreadcrumbsContext = createContext<BreadcrumbsState>(new BreadcrumbsState([]));
