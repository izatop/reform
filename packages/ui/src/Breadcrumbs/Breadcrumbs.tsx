import {Breadcrumbs as BC} from "@reform/bulma";
import * as React from "react";
import {useContext} from "react";
import {BreadcrumbLink} from "./BreadcrumbLink";
import {BreadcrumbsContext} from "./BreadcrumbsProvider";

export type Breadcrumb = (path: string, children: React.ReactElement) => React.ReactElement;

export interface BreadcrumbsProps {
    children?: Breadcrumb;
}

const defaultChildRender: Breadcrumb = (path, children) => (<a href={path}>{children}</a>);

export const Breadcrumbs: React.FC<BreadcrumbsProps> = ({children = defaultChildRender}) => {
    const breadcrumbs = useContext(BreadcrumbsContext);
    const paths = breadcrumbs.usePaths();

    return (
        <BC>
            {paths.map(([path, title, icon], i) => (
                <React.Fragment key={i}>
                    {children(path, <BreadcrumbLink icon={icon} title={title}/>)}
                </React.Fragment>
            ))}
        </BC>
    );
};
