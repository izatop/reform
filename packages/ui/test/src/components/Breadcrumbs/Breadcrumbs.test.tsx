import * as React from "react";
import * as renderer from "react-test-renderer";
import {Breadcrumbs, BreadcrumbsProvider} from "../../../../src/Breadcrumbs";

test("Non ideal state", () => {
    const body = (
        <BreadcrumbsProvider base={[["/", "Home", "icon"], ["/me", "User"]]}>
            <Breadcrumbs/>
        </BreadcrumbsProvider>
    );

    expect(renderer.create(body).toJSON())
        .toMatchSnapshot();
});
