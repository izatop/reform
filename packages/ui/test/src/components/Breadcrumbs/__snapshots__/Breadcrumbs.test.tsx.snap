// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Non ideal state 1`] = `
<nav
  aria-label="breadcrumbs"
  className="breadcrumb"
>
  <ul>
    <li
      className=""
    >
      <a
        href="/"
      >
        <span
          className="icon"
        >
          <i
            className="fas fa-icon"
          />
        </span>
        <span>
          Home
        </span>
      </a>
    </li>
    <li
      className="is-active"
    >
      <a
        href="/me"
      >
        User
      </a>
    </li>
  </ul>
</nav>
`;
